ow_power_system/src/PowerSystemNode.cpp

```diff
@@ -401,167 +455,369 @@ void PowerSystemNode::injectCustomFault(bool& fault_activated,
 }
 
-void PowerSystemNode::Run()
+/*
+ * Attempts to call loadPowerProfile and returns true or false depending on its success.
+ */
+bool PowerSystemNode::loadCustomFaultPowerProfile(std::string path, std::string custom_file)
 {
-  ROS_INFO("Power system node running.");
-
-  // Start the asynchronous spinner, which will call jointStatesCb as the
-  // joint_states topic is published (50Hz).
-  ros::AsyncSpinner spinner(m_spinner_threads);
-  spinner.start();
+  m_custom_power_fault_sequence = loadPowerProfile(path, custom_file);
 
-  // For simplicity, we run the power node at the same rate as GSAP.
-  ros::Rate rate(m_gsap_rate_hz);
+  // Return false if the sequence was not properly initialized.
+  return (m_custom_power_fault_sequence.size() > 0);
+}
 
-  while (ros::ok())
+/*
+ * Called at the end of a cycle, this publishes RUL/SoC/battery temperature
+ * based on all prognoser predictions obtained.
+ */
+void PowerSystemNode::publishPredictions()
+{
+  // Using EoD_events, publish the relevant values.
+
+  // Start the values at either their maximum possible, or beyond the minimum.
+  double min_rul = m_max_horizon_secs;
+  double min_soc = 1;
+  double max_tmp = -1000;
+  owl_msgs::BatteryRemainingUsefulLife rul_msg;
+  owl_msgs::BatteryStateOfCharge soc_msg;
+  owl_msgs::BatteryTemperature tmp_msg;
+
+  // Calculate the min_rul, min_soc, and max_tmp from the new data.
+  for (int i = 0; i < NUM_MODELS; i++)
   {
-    // If there is a new mechanical power average ready to send to GSAP, do so.
-    if (m_trigger_processing_new_power_batch)
+    // If RUL is infinity, set it to the maximum horizon value instead.
+    if (isinf(m_EoD_events[i].remaining_useful_life))
     {
-      m_trigger_processing_new_power_batch = false;
-      m_processing_power_batch = true;
-      runPrognoser(m_mechanical_power_to_be_processed / m_efficiency);
-      m_processing_power_batch = false;
+      m_EoD_events[i].remaining_useful_life = m_max_horizon_secs;
     }
 
-    rate.sleep();
+    min_rul = std::min(min_rul, m_EoD_events[i].remaining_useful_life);
+    min_soc = std::min(min_soc, m_EoD_events[i].state_of_charge);
+    max_tmp = std::max(max_tmp, m_EoD_events[i].battery_temperature);
   }
 
-  spinner.stop();
+  // DEBUG PRINT
+  if (m_topics_print_debug)
+  {
+    printTopics(min_rul, min_soc, max_tmp);
+  }
+
+  // If either RUL or SoC have entered the negative, it indicates the battery
+  // is in a fail state.
+  // While it is possible sometimes for the battery to recover from this,
+  // differentiating between the 2 different types of fail states has not been
+  // implemented yet.
+  // For now, treat all fail states as permanent and flatline all future
+  // publications.
+  // The timestamp stipulation is to prevent odd/invalid values during startup
+  // from triggering the fail state.
+  if ((min_rul < 0 || min_soc < 0)
+      && (!m_battery_failed)
+      && (m_power_models[0].input_info.timestamp >= (m_time_interval * 5)))
```

Comment:
Yeah, if i recall correctly, the power loop seems to cycle at least once before the simulation starts running properly and during that time all of the values are 0. If there isn't a catch to ignore these starting values, the fail state will immediately trigger on launch. The 5 is an arbitrary value I set to give some leeway just in case. It could be adjusted, or if you've got a better idea for handling it I'm all ears.

Otherwise, maybe I should at least label it as a hack. It's not exactly addressing the base problem of the loop running before the simulation has stabilized.

Reply:
At the very least a comment would be good. The only alternative I can think of is to run a loop ahead of this for however many seconds (e.,g. 5) and throw out the values, then run this loop.

But what you have here accomplishes the same thing fewer less lines of code, so I think the comment is sufficient.